### ISSUE TITLE
Post Object Restore

### ISSUE THREAD

**Issue by geoffreak** [chilts/awssum-amazon-s3#1]
Since you requested I make a ticket here, here it is. I actually need this functionality now, so I went ahead and added this to the existing `s3-config.js`, but I have yet to test it. If you notice any bugs, please let me know.

```
    // Post Object Restore

    PostObjectRestore : {
        url : 'http://docs.amazonwebservices.com/AmazonS3/latest/API/RESTObjectPOSTrestore.html',
        // request
        method : 'POST',
        host   : hostBucket,
        path   : pathObject,
        defaults : {
            restore : undefined,
        },
        args : {
            BucketName : {
                required : true,
                type     : 'special',
            },
            restore : {
                required : true,
                type     : 'resource',
            },
            Body : {
                required : true,
                type     : 'body',
            },
        },
        addExtras : extrasContentMd5,
        // response
        extractBody : 'none',
    },

```

**Comment by chilts** [comment]
Hey Joshua,

That's lovely, thanks. :) Since you've added it (and I'm now in my working day), I'll let you carry on for now and I'll check it in later on today.

Awesome, thanks again.

Cheers,
Andy

**Comment by chilts** [comment]
Wasn't quite as straightforward as you suggested, just one thing missing (the generation of the body). :)

Published to npm as v1.2.1.

Thanks again,
Andy

**Comment by chilts** [comment]
Ah crap. Re-published as v1.2.2 !!!

:+1: :cactus: :dancers: :earth_americas:  etc etc :)

**Comment by geoffreak** [comment]
It doesn't appear that the MD5 signature is being calculated properly. I'm getting the following error back from S3:

```
Code: 'SignatureDoesNotMatch',
Message: 'The request signature we calculated does not match the signature you provided. Check your key and signing method.',

```

**Comment by geoffreak** [comment]
Ah, looks like the following XML is being generated:

``` xml
<RestoreRequest><@><xmlns>http://s3.amazonaws.com/doc/2006-03-01/</xmlns></@><Days>1</Days></RestoreRequest>
```

**Comment by geoffreak** [comment]
Even after manually correcting the xml, I still cannot get the signature to match. Ugh.

```
-------------------------------------------------------------------------------
Request:
- method         :  POST
- protocol       :  https
- host           :  archive.s3.amazonaws.com
- path           :  /cli.zip
- params         :  [ { name: 'restore', value: undefined } ]
- headers        :  { 'Content-MD5': '94c8afd21410a4191d3fedcfc31d1db2',
  Date: 'Wed, 27 Mar 2013 16:19:29 UTC',
  Authorization: 'AWS **************************************',
  'User-Agent': 'awssum/0.12.2' }
- forms          :  []
- json           :  {}
- body           :  <RestoreRequest xmlns="http://s3.amazonaws.com/doc/2006-3-01"><Days>1</Days></RestoreRequest>
Request:
- statusCode     : 200
- extractHeaders : true
- extractBody : none
-------------------------------------------------------------------------------
```

**Comment by chilts** [comment]
ok thanks. I'll look into it shortly :)

**Comment by chilts** [comment]
I've tried everything and it's not working. Every single other operation in S3 _is_ working, so either their docs are wrong, or that operation is currently broken (I suspect the former).

I have followed up on their forums with some clarification for someone posting this exact problem using AwsSum:
- https://forums.aws.amazon.com/thread.jspa?messageID=437285&#437285

Therefore, I'm not going to put any extra time into this until we get a response from them. I hope that's ok. :)

Cheers,
Andy

**Comment by geoffreak** [comment]
I see you found the ticket we made on the developer forums (Dan is the one at our company with the AWS account). Hopefully someone will come along and figure out why it isn't working. I realize now that I accidentally had a non-base64 md5 in the ticket, but your clarification corrects that.

**Comment by geoffreak** [comment]
I just used the PHP AWS library to make a valid restoreObject request and here's the (working!) HTTP request it made:

``` http
POST /cli.zip?restore HTTP/1.1
Host: searchtower-archive.s3.amazonaws.com
User-Agent: aws-sdk-php2/2.2.1 Guzzle/3.3.1 curl/7.21.4 PHP/5.3.15
Content-Length: 118
Content-Type: application/xml
Date: Thu, 28 Mar 2013 14:42:02 +0000
Authorization: AWS ************

<?xml version="1.0"?>
<RestoreRequest xmlns="http://s3.amazonaws.com/doc/2006-03-01/"><Days>1</Days></RestoreRequest>
```

and the request coming from awssum (modified by me to match the above, but still not working):

``` http
POST /cli.zip?restore HTTP/1.1
Content-Length: 134
Content-Type: application/xml
Date: Thu, 28 Mar 2013 15:12:12 UTC
Authorization: AWS *************
User-Agent: awssum/0.12.2
Host: searchtower-archive.s3.amazonaws.com
Connection: keep-alive

<?xml version="1.0" encoding="utf-8"?>
<RestoreRequest xmlns="http://s3.amazonaws.com/doc/2006-03-01/"><Days>1</Days></RestoreRequest>
```

Another thing to note: PHP uses https whereas awssum uses http. It doesn't appear to make a difference when I switch awssum to only use https though.

**Comment by chilts** [comment]
Hmm, you shouldn't have awssum/0.12.2 in your User-Agent. Can you get rid of that and install awssum-amazon-s3 (which will install AwsSum v1.0.0-beta) and use that instead.

Not sure why you said AwsSum uses http, it's always been https. In fact, I've only recently added the ability to switch to plain http. :)

The main thing I see is that they are not providing the Content-MD5 header so I'm taking that out (for now) and trying again.

Thanks for your help,
Andy

**Comment by chilts** [comment]
@geoffreak 

Hi Joshua,

I've just fixed this and pushed v1.2.3 to npm. Please download it and have a try! :) Hopefully you're all sorted now.

I've also replied here:
- https://forums.aws.amazon.com/thread.jspa?messageID=437536&#437536

Thanks for raising the issue, helping debug it and being patient in my crazy world. :)

Cheers,
Andy

**Comment by geoffreak** [comment]
Thanks for all the help!
